math/mathcore/inc/Math/Functor.h

```diff
@@ -741,62 +181,50 @@ class GradFunctor : public IGradientFunctionMultiDim  {
     * @param dim : number of function dimension
     * @param g   : function object computing the function gradient
     */
-   GradFunctor(const std::function<double(double const *)> &f, int dim,
-               const std::function<void(double const *, double *)> &g)
-      : fImpl(new FunctorGradHandler<GradFunctor>(dim, f, g))
+   GradFunctor(std::function<double(double const *)> const&f, unsigned int dim,
+               std::function<void(double const *, double *)> const& g)
+      : fDim{dim}, fFunc{f}, fGradFunc{g}
    {}
 
-   /**
-      Destructor (no operations)
-   */
-   virtual ~GradFunctor ()  {}
-
-
-   /**
-      Copy constructor for functor based on ROOT::Math::IMultiGradFunction
-   */
-   GradFunctor(const GradFunctor & rhs) :
-      ImplBase()
-   {
-      if (rhs.fImpl)
-         fImpl = std::unique_ptr<Impl>(rhs.fImpl->Copy());
-   }
-
-   /**
-      Assignment operator
-   */
-   GradFunctor & operator = (const GradFunctor & rhs)  {
-      GradFunctor copy(rhs);
-      fImpl.swap(copy.fImpl);
-      return *this;
-   }
-
-
-   // clone of the function handler (use copy-ctor)
-   ImplBase * Clone() const { return new GradFunctor(*this); }
+   // Clone of the function handler (use copy-ctor).
+   GradFunctor * Clone() const { return new GradFunctor(*this); }
 
    // for multi-dimensional functions
-   unsigned int NDim() const { return fImpl->NDim(); }
+   unsigned int NDim() const { return fDim; }
 
    void Gradient(const double *x, double *g) const {
-      fImpl->Gradient(x,g);
+      // Fall back to base implementation if no gradient function is provided
+      // (it will fill the gradient calling DoDerivative() for each component).
+      if(!fGradFunc) {
+         IGradientFunctionMultiDim::Gradient(x, g);
+         return;
+      }
+      fGradFunc(x, g);
    }
 
 private :
 
-
    inline double DoEval (const double * x) const {
-      return (*fImpl)(x);
+      return fFunc(x);
    }
 
-
    inline double DoDerivative (const double * x, unsigned int icoord  ) const {
-      return fImpl->Derivative(x,icoord);
+      if(fDerivFunc) {
+         return fDerivFunc(x, icoord);
+      }
+      // Get the component from the gradient if not derivative function is
+      // specified.
+      gradBuffer.resize(fDim);
+      std::fill(gradBuffer.begin(), gradBuffer.end(), 0.0);
+      fGradFunc(x, gradBuffer.data());
+      return gradBuffer[icoord];
    }
 
-   std::unique_ptr<Impl> fImpl;    // pointer to base grad functor handler
-
-
+   unsigned int fDim;
+   std::function<double(const double *)> fFunc;
+   std::function<double(double const *, unsigned int)> fDerivFunc;
+   std::function<void(const double *, double*)> fGradFunc;
+   mutable std::vector<double> gradBuffer;
```

Comment:
I think is not worth adding this mutable data member and make the function not able to work anymore in multi-threads. 
If the function is complex allocating a buffer is probably ok, and if one wants optimize code it should not use this function when fDerivFunc is not set.

Reply:
I agree with you, this little optimization is probably not worth introducing a `mutable` member for, given that `DoDerivative` should not be called anyway if you have the full gradient